File: java/src/main/java/io/hron/ast/DayFilter.java
package io.hron.ast;

import java.util.List;

/**
 * Represents a filter for which days a schedule applies to.
 *
 * @param kind the type of day filter
 * @param days the specific days (only used when kind is DAYS)
 */
public record DayFilter(Kind kind, List<Weekday> days) {

  /** The type of day filter. */
  public enum Kind {
    /** Matches every day. */
    EVERY,
    /** Matches weekdays (Monday-Friday). */
    WEEKDAY,
    /** Matches weekend days (Saturday-Sunday). */
    WEEKEND,
    /** Matches specific days of the week. */
    DAYS
  }

  /**
   * Creates a filter that matches every day.
   *
   * @return a new every-day filter
   */
  public static DayFilter every() {
    return new DayFilter(Kind.EVERY, List.of());
  }

  /**
   * Creates a filter that matches weekdays (Mon-Fri).
   *
   * @return a new weekday filter
   */
  public static DayFilter weekday() {
    return new DayFilter(Kind.WEEKDAY, List.of());
  }

  /**
   * Creates a filter that matches weekends (Sat-Sun).
   *
   * @return a new weekend filter
   */
  public static DayFilter weekend() {
    return new DayFilter(Kind.WEEKEND, List.of());
  }

  /**
   * Creates a filter that matches specific days.
   *
   * @param days the days to match
   * @return a new specific-days filter
   */
  public static DayFilter days(List<Weekday> days) {
    return new DayFilter(Kind.DAYS, List.copyOf(days));
  }
}

File: csharp/Hron/Ast/DayFilter.cs
namespace Hron.Ast;

/// <summary>
/// Represents a filter for which days a schedule applies to.
/// </summary>
public sealed record DayFilter(DayFilterKind Kind, IReadOnlyList<Weekday> Days)
{
    /// <summary>
    /// Creates a filter that matches every day.
    /// </summary>
    public static DayFilter Every() => new(DayFilterKind.Every, []);

    /// <summary>
    /// Creates a filter that matches weekdays (Mon-Fri).
    /// </summary>
    public static DayFilter Weekday() => new(DayFilterKind.Weekday, []);

    /// <summary>
    /// Creates a filter that matches weekends (Sat-Sun).
    /// </summary>
    public static DayFilter Weekend() => new(DayFilterKind.Weekend, []);

    /// <summary>
    /// Creates a filter that matches specific days.
    /// </summary>
    public static DayFilter SpecificDays(IReadOnlyList<Weekday> days) => new(DayFilterKind.Days, days);
}

/// <summary>
/// The type of day filter.
/// </summary>
public enum DayFilterKind
{
    /// <summary>Matches every day.</summary>
    Every,
    /// <summary>Matches weekdays (Monday-Friday).</summary>
    Weekday,
    /// <summary>Matches weekend days (Saturday-Sunday).</summary>
    Weekend,
    /// <summary>Matches specific days of the week.</summary>
    Days
}
